### PR TITLE
feat: change sync mode to block by block

### PR DESCRIPTION
**Pull request by shaojunda**

### CODE REVIEW

File: app/models/block.rb [hunk 1]
```diff
@@ -25,21 +25,12 @@ class Block < ApplicationRecord
   attribute :proposals, :ckb_array_hash, hash_length: ENV["DEFAULT_SHORT_HASH_LENGTH"]
 
   scope :recent, -> { order(timestamp: :desc) }
-  scope :available, -> { where(status: [:inauthentic, :authentic]) }
+  scope :accepted, -> { where(status: :accepted) }
```
Comment (quake) [inline app/models/block.rb]: ActiveRecord provides scope based on the values of the enum field, we can remove this line.

File: app/models/block.rb [hunk 2]
```diff
@@ -88,25 +75,10 @@ def flush_cache
...
-    self
-  end
-
-  def abandon!
-    update!(status: "abandoned", reward_status: "issued")
-    uncle_blocks.update_all(status: "abandoned")
-    ChangeCkbTransactionsStatusWorker.perform_async(id, "abandoned")
-    ChangeCellOutputsStatusWorker.perform_async(id, "abandoned")
+  def invalid!
+    update!(status: "abandoned")
```
Comment (quake) [inline app/models/block.rb]:
```suggestion
    abandoned!
```

File: app/models/cell_output.rb [hunk 3]
```diff
@@ -1,19 +1,19 @@
...
-  has_one :type_script
+  has_one :lock_script, dependent: :delete
+  has_one :type_script, dependent: :delete
 
   validates :capacity, presence: true, numericality: { greater_than_or_equal_to: 0 }
 
   attribute :tx_hash, :ckb_hash
 
-  scope :available, -> { where(status: [:live, :dead]) }
+  scope :accepted, -> { where(status: [:live, :dead]) }
```
Comment (quake) [inline app/models/cell_output.rb]: useless scope, it's equal to `all`

File: app/models/ckb_sync/node_data_processor.rb [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+        proposals_count: uncle_block.proposals.count,
+        witnesses_root: header.witnesses_root,
+        epoch: header.epoch,
+        dao: header.dao
+      )
+    end
+
+    def build_ckb_transactions(local_block, transactions)
+      transaction_index = 0
+      transactions.map do |transaction|
```
Comment (quake) [inline app/models/ckb_sync/node_data_processor.rb]:
```suggestion
      transactions.each_with_index.map do |transaction, transaction_index|
```

File: app/models/ckb_sync/node_data_processor.rb [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        previous_output: node_input.previous_output,
+        since: node_input.since,
+        block: ckb_transaction.block,
+        from_cell_base: cell.blank?
+      )
+    end
+
+    def build_cell_outputs(node_outputs, ckb_transaction, addresses)
+      cell_index = 0
+      node_outputs.map do |output|
```
Comment (quake) [inline app/models/ckb_sync/node_data_processor.rb]: please use `each_with_index.map`

File: app/models/miner_ranking.rb [hunk 6]
```diff
@@ -26,7 +26,7 @@ def ranking_infos
     addresses = Address.where(id: address_ids)
     ranking_infos = []
     addresses.find_each do |address|
-      block_ids = address.ckb_transactions.available.where(is_cellbase: true).select("block_id")
+      block_ids = address.ckb_transactions.where(is_cellbase: true).select("block_id")
```
Comment (quake) [inline app/models/miner_ranking.rb]:
```suggestion
      block_ids = address.ckb_transactions.where(is_cellbase: true).pluck("block_id")
```

File: app/models/ckb_sync/node_data_processor.rb (hunk 4 above)
Comment (shaojunda) [reply app/models/ckb_sync/node_data_processor.rb]: https://gist.github.com/spider-network/8b0e168a932731eade6f `each` may faster than `each_with_index`

File: app/models/miner_ranking.rb (hunk 6 above)
Comment (shaojunda) [reply app/models/miner_ranking.rb]: `select` will return relation. `pluck` will return array. if `select` reduce the amount of data the query？

File: app/models/block.rb (hunk 1 above)
Comment (shaojunda) [reply app/models/block.rb]: OK，I will fix this.

File: app/models/cell_output.rb (hunk 3 above)
Comment (shaojunda) [reply app/models/cell_output.rb]: ok, I will fix this.

File: app/models/ckb_sync/node_data_processor.rb (hunk 4 above)
Comment (shaojunda) [reply app/models/ckb_sync/node_data_processor.rb]: https://media.pragprog.com/titles/adrpo/iterators.pdf this show me  `each_with_index` will create  additional T_NODE. so I choose to use each.

File: app/models/ckb_sync/node_data_processor.rb (hunk 5 above)
Comment (shaojunda) [reply app/models/ckb_sync/node_data_processor.rb]: thank you. fixed.

**Pull request merged**